baseplate/experiments/__init__.py

```diff
@@ -0,0 +1,171 @@
+        self._already_bucketed = set()
+
+    def _get_config(self, name):
+        try:
+            config_data = self._config_watcher.get_data()
+            return config_data[name]
+        except WatchedFileNotAvailableError:
+            logger.warning("Experiment config file not found")
+            return
+        except KeyError:
+            logger.warning(
+                "Experiment <%r> not found in experiment config",
+                name,
+            )
+            return
+        except TypeError:
+            logger.warning("Could not load experiment config: %r", name)
+            return
+
+    def variant(self, name, bucketing_event_override=None,
+                extra_event_params=None, **kwargs):
+        """ Which variant, if any, is active for the experiment specified by
+        "name".  If a variant is active, a bucketing event will be logged to
+        the event pipeline unless any one of the following conditions are met:
+
+        1. bucketing_event_override is set to False.
+        2. The experiment specified by "name" explicitly disables bucketing
+           events.
+        3. We have already logged a bucketing event for the value specified by
+           experiment.event_cache_key(**kwargs) within the current session.
+
+        :param str name: Name of the experiment you want to run.
+        :param bool bucketing_event_override: Optional value, defaults to None.
+            If set to True, will always log bucketing events unless the
+            experiment explicitly disables them.  If set to False, will never
+            send a bucketing event.  If set to None, no override will be
+            applied.
+        :param dict extra_event_params: Optional value.  Any extra values you
+            want to add to the bucketing event.
+        :param kwargs:  Arguments that will be passed to experiment.variant to
+            determine bucketing, targeting, and overrides.
+
+        :rtype: :py:class:`str`
+        :return: Variant name if a variant is active, None otherwise.
+        """
+        config = self._get_config(name)
+        if not config:
+            return None
+
+        experiment = parse_experiment(config)
+        variant = experiment.variant(**kwargs)
+
+        do_log = True
+
+        if variant is None:
+            do_log = False
+
+        cache_key = experiment.event_cache_key(**kwargs)
+
+        if cache_key and cache_key in self._already_bucketed:
+            do_log = False
+
+        if bucketing_event_override is True:
+            do_log = True
+        elif bucketing_event_override is False:
+            do_log = False
+
+        do_log = do_log and experiment.should_log_bucketing()
+
+        if do_log:
+            self._log_bucketing_event(experiment, variant, extra_event_params)
+            if cache_key:
+                self._already_bucketed.add(cache_key)
+
+        return variant
+
+    def _log_bucketing_event(self, experiment, variant, extra_event_params=None):
```

Comment:
does `extra_event_params` need to take a default since it's only used internally once?